subsys/net/lib/zoap/zoap.c

```diff
@@ -615,19 +615,23 @@ struct zoap_reply *zoap_response_received(
 {
 	struct zoap_reply *r;
 	const u8_t *token;
+	u16_t id;
 	u8_t tkl;
 	size_t i;
 
+	id = zoap_header_get_id(response);
 	token = zoap_header_get_token(response, &tkl);
 
 	for (i = 0, r = replies; i < len; i++, r++) {
 		int age;
 
-		if (r->tkl != tkl) {
+		if ((r->id == 0) &&
```

Comment:
Apparently there's some more testing to do with this patch.  It might cause a regression where we're matching erroneously.  This patch fixes a RST return from Leshan on an observer notification where the user as stopped the observer.  The RST message has an ID which matches the newly stored reply->id, but there's no referencing token.  I should probably add this information to the patch description.